lib/pengine/rules.c

```diff
@@ -965,12 +965,163 @@ pe_eval_subexpr(xmlNode *expr, pe_rule_eval_data_t *rule_data, crm_time_t *next_
+            int rc1 = errno;
+
+            long long r_val_num = crm_parse_ll(r_val, NULL);
+            int rc2 = errno;
+
+            if (rc1 == 0 && rc2 == 0) {
+                if (l_val_num < r_val_num) {
+                    cmp = -1;
+                } else if (l_val_num > r_val_num) {
+                    cmp = 1;
+                } else {
+                    cmp = 0;
+                }
+            } else {
+                crm_debug("Integer parse error. Comparing %s and %s as strings",
+                          l_val, r_val);
+                cmp = compare_attr_expr_vals(l_val, r_val, "string", op);
+            }
+
+        } else if (pcmk__str_eq(type, "number", pcmk__str_casei)) {
+            double l_val_num;
+            double r_val_num;
+
+            int rc1 = pcmk__scan_double(l_val, &l_val_num, NULL, NULL);
+            int rc2 = pcmk__scan_double(r_val, &r_val_num, NULL, NULL);
+
+            if (rc1 == pcmk_rc_ok && rc2 == pcmk_rc_ok) {
+                if (l_val_num < r_val_num) {
+                    cmp = -1;
+                } else if (l_val_num > r_val_num) {
+                    cmp = 1;
+                } else {
+                    cmp = 0;
+                }
+
+            } else {
+                crm_debug("Floating-point parse error. Comparing %s and %s as "
+                          "strings", l_val, r_val);
+                cmp = compare_attr_expr_vals(l_val, r_val, "string", op);
+            }
+
+        } else if (pcmk__str_eq(type, "version", pcmk__str_casei)) {
+            cmp = compare_version(l_val, r_val);
+
+        }
+
+    } else if (l_val == NULL && r_val == NULL) {
+        cmp = 0;
+    } else if (r_val == NULL) {
+        cmp = 1;
+    } else {    // l_val == NULL && r_val != NULL
+        cmp = -1;
+    }
+
+    return cmp;
+}
+
+/*!
+ * \internal
+ * \brief   Check whether an attribute expression evaluates to \c true
+ *
+ * \param[in]   l_val   Value on left-hand side of comparison
+ * \param[in]   r_val   Value on right-hand side of comparison
+ * \param[in]   type    How to interpret the values (allowed values:
+ *                      "string", "integer", "number", "version",
+ *                      \c NULL)
+ * \param[in]   op      Type of comparison.
+ *
+ * \return  \c true if expression evaluates to \c true, \c false
+ *          otherwise
+ */
+static bool
+accept_attr_expr(const char *l_val, const char *r_val, const char *type,
+                 const char *op)
+{
+    int cmp;
+
+    if (pcmk__str_eq(op, "defined", pcmk__str_casei)) {
+        return (l_val != NULL);
+
+    } else if (pcmk__str_eq(op, "not_defined", pcmk__str_casei)) {
+        return (l_val == NULL);
+    }
+
+    cmp = compare_attr_expr_vals(l_val, r_val, type, op);
+
+    if (pcmk__str_eq(op, "eq", pcmk__str_casei)) {
+        return (cmp == 0);
+
+    } else if (pcmk__str_eq(op, "ne", pcmk__str_casei)) {
+        return (cmp != 0);
+
+    } else if (l_val == NULL || r_val == NULL) {
```

Comment:
It seems like we could do without this. At best, it saves us from four string comparisons that follow this block.

compare_attr_expr_vals() treats a NULL value as 0 or an empty string for the sake of comparison. If we eliminate this check, then we just use that result from compare_attr_expr_vals() instead of saying "the whole expression evaluates to false no matter what." It would be a slight change in logic, but the existing behavior was not publicly defined.

Reply:
> compare_attr_expr_vals() treats a NULL value as 0 or an empty string for the sake of comparison.

Actually NULL always compares as less than non-NULL (the comparison value could be negative).



>  It would be a slight change in logic, but the existing behavior was not publicly defined.

Yes but it's better not to surprise people on upgrades. :)

I think the current approach makes sense from the user's point of view anyway. If a user wants some configuration "when the value of node attribute A is less than some number N", I'm guessing they want that to be considered true only if there is an actual attribute A specified for the node. We'd need a new "value-default" attribute for <expression> to let the user decide what "no attribute exists" should mean, if we wanted to change that. (Which of course we don't want to do unless there is demand for it.)

Reply:
> Actually NULL always compares as less than non-NULL (the comparison value could be negative).

I honestly don't know why I typed that it's treated as zero.. You're right of course and I saw that too.